[dtensor][partial] improving followed_strategy selection when no sharding placement for min/max

anshul-si opened the pull request:
**Summary:** The current problem is when we have no sharding placements in any of the input_args, we choose first arg's strategy as followed_strategy in pointwise_ops, which can lead to unnecessary redistributes. One example of this is torch.maximum(p(max), p(sum)) vs torch.maximum(p(sum), p(max)). In the first instance, we choose p(max) and propagate it. p(sum) is then converted to replicate. Overall we do one redistribute. However, for the second case, we choose p(sum) to propagate. We choose replicate as linearity is -1, and we check if sum == max. This results in both args being converted to Replicate. 

I solve this issue by first checking if any of the args possible placements contain Shard(). If they don't, for every argument, we calculate the index of the last Partial placement that doesn't match the preserved reduce_op (e.g., "max" for torch.maximum). This is because for this placement, we will have to redistribute to Replicate, and in a correct implementation, we would have to Replicate all placements that come before it (including Partials that do match the reduce_op). In summary choosing the argument where the last non-matching Partials appear earliest in the placement list minimizes the number of mesh dimensions requiring redistribution. 

**Design Choice/Question**
1. Converted partial_preserving_ops to map to list instead of strings as we can convert linear_ops to follow this format, and some ops can propagate multiple types of partials.

2. Currently I check every arg before hand if it contains a spec with Shard. The reason I can't just use max_shard to tell me this is because the max_num_shards() returns atleast 1 no matter what (even with no sharding) However, there is, what i think is an extreme edge case where if the mesh_size is 1 and there's one shard placement, we return 1. If we accept that in this instance, we're ok with converting to Partial sometimes, i can remove logic checking the arg for Shard and just do if preserve_partial is not None and max_shards == 1:

**Test Cases**
1. pytest test/distributed/tensor/test_pointwise_ops.py -k test_maximum_mixed_partials_redistribution
2. pytest test/distributed/tensor/test_pointwise_ops.py -k test_maximum_selects_optimal_strategy

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #170632
* #170040
* #170035
* #170030



cc @wanchaol @tianyu-l @wz337 @XilunWu @d4l3k @pragupta @SherlockNoMad @ppwwyyxx

anshul-si committed:
[dtensor][partial] improving followed_strategy selection when no sharding

[ghstack-poisoned]

anshul-si committed:
Update on "[dtensor][partial] improving followed_strategy selection when no sharding placement for min/max"




cc wanchaol tianyu-l wz337 XilunWu d4l3k pragupta SherlockNoMad ppwwyyxx

[ghstack-poisoned]

wconstab commented on torch/distributed/tensor/_ops/_pointwise_ops.py:
```diff
@@ -427,22 +427,60 @@
-    aten.maximum.out: "max",
-    aten.minimum.default: "min",
-    aten.minimum.out: "min",
+partial_preserving_ops: dict[torch._ops.OpOverload, list[str]] = {
```
IIUC you are changing this type with the intent of supporting other ops?  Can you give a prediction of what this list would look like in the future and how it would impact the rest of the pointwise design?   Specifically, do we delete 'linearity' and rely solely on op:[reduce ops] mappings? I guess what i'm asking for is not really a PR comment or change, but separately, some kind of a design proposal that says what the metadata format will look like and what the logic will look like.

anshul-si replied:
Yes, was hoping to completely remove linearity. I did it for just maximum and minimum since we wanted a fix by Wednesday

wconstab commented on torch/distributed/tensor/_ops/_pointwise_ops.py:
```diff
@@ -427,22 +427,60 @@
+    Returns the minimum "last non-matching partial index" across all OpSpecs.
+
+    For each OpSpec, finds the index of the last Partial that doesn't match
+    preserve_partial_types. Takes the min across all OpSpecs (best case).
```
an example would help here.

Also, you might be the first person to talk about 'Partial order' so it bears writing something explicitly about what your assumptions are here.  We are allowing shards to be in any order (using ShardOrder) but are we assuming partials are always created in L2R order?  How does DTensor currently control the creation order of Partials?

anshul-si committed:
Update on "[dtensor][partial] improving followed_strategy selection when no sharding placement for min/max"

**Summary:** The current problem is when we have no sharding placements in any of the input_args, we choose first arg's strategy as followed_strategy in pointwise_ops, which can lead to unnecessary redistributes. One example of this is torch.maximum(p(max), p(sum)) vs torch.maximum(p(sum), p(max)). In the first instance, we choose p(max) and propagate it. p(sum) is then converted to replicate. Overall we do one redistribute. However, for the second case, we choose p(sum) to propagate. We choose replicate as linearity is -1, and we check if sum == max. This results in both args being converted to Replicate. 

I solve this issue by first checking if any of the args possible placements contain Shard(). If they don't, for every argument, we calculate the index of the last Partial placement that doesn't match the preserved reduce_op (e.g., "max" for torch.maximum). This is because for this placement, we will have to redistribute to Replicate, and in a correct implementation, we would have to Replicate all placements that come before it (including Partials that do match the reduce_op). In summary choosing the argument where matching Partials appear earlier in the placement list minimizes the number of mesh dimensions requiring redistribution. 

**Design Choice/Question**
1. Converted partial_preserving_ops to map to list instead of strings as we can convert linear_ops to follow this format, and some ops can propagate multiple types of partials.

2. Currently I check every arg before hand if it contains a spec with Shard. The reason I can't just use max_shard to tell me this is because the max_num_shards() returns atleast 1 no matter what (even with no sharding) However, there is, what i think is an extreme edge case where if the mesh_size is 1 and there's one shard placement, we return 1. If we accept that in this instance, we're ok with converting to Partial sometimes, i can remove logic checking the arg for Shard and just do if preserve_partial is not None and max_shards == 1:

**Test Cases**
1. pytest test/distributed/tensor/test_pointwise_ops.py -k test_maximum_mixed_partials_redistribution
2. pytest test/distributed/tensor/test_pointwise_ops.py -k test_maximum_selects_optimal_strategy




cc wanchaol tianyu-l wz337 XilunWu d4l3k pragupta SherlockNoMad ppwwyyxx

[ghstack-poisoned]

anshul-si committed:
Update on "[dtensor][partial] improving followed_strategy selection when no sharding placement for min/max"

**Summary:** The current problem is when we have no sharding placements in any of the input_args, we choose first arg's strategy as followed_strategy in pointwise_ops, which can lead to unnecessary redistributes. One example of this is torch.maximum(p(max), p(sum)) vs torch.maximum(p(sum), p(max)). In the first instance, we choose p(max) and propagate it. p(sum) is then converted to replicate. Overall we do one redistribute. However, for the second case, we choose p(sum) to propagate. We choose replicate as linearity is -1, and we check if sum == max. This results in both args being converted to Replicate. 

I solve this issue by first checking if any of the args possible placements contain Shard(). If they don't, for every argument, we calculate the index of the last Partial placement that doesn't match the preserved reduce_op (e.g., "max" for torch.maximum). This is because for this placement, we will have to redistribute to Replicate, and in a correct implementation, we would have to Replicate all placements that come before it (including Partials that do match the reduce_op). In summary choosing the argument where matching Partials appear earlier in the placement list minimizes the number of mesh dimensions requiring redistribution. 

**Design Choice/Question**
1. Converted partial_preserving_ops to map to list instead of strings as we can convert linear_ops to follow this format, and some ops can propagate multiple types of partials.

2. Currently I check every arg before hand if it contains a spec with Shard. The reason I can't just use max_shard to tell me this is because the max_num_shards() returns atleast 1 no matter what (even with no sharding) However, there is, what i think is an extreme edge case where if the mesh_size is 1 and there's one shard placement, we return 1. If we accept that in this instance, we're ok with converting to Partial sometimes, i can remove logic checking the arg for Shard and just do if preserve_partial is not None and max_shards == 1:

**Test Cases**
1. pytest test/distributed/tensor/test_pointwise_ops.py -k test_maximum_mixed_partials_redistribution
2. pytest test/distributed/tensor/test_pointwise_ops.py -k test_maximum_selects_optimal_strategy




cc wanchaol tianyu-l wz337 XilunWu d4l3k pragupta SherlockNoMad ppwwyyxx

[ghstack-poisoned]

anshul-si committed:
Update on "[dtensor][partial] improving followed_strategy selection when no sharding placement for min/max"

**Summary:** The current problem is when we have no sharding placements in any of the input_args, we choose first arg's strategy as followed_strategy in pointwise_ops, which can lead to unnecessary redistributes. One example of this is torch.maximum(p(max), p(sum)) vs torch.maximum(p(sum), p(max)). In the first instance, we choose p(max) and propagate it. p(sum) is then converted to replicate. Overall we do one redistribute. However, for the second case, we choose p(sum) to propagate. We choose replicate as linearity is -1, and we check if sum == max. This results in both args being converted to Replicate. 

I solve this issue by first checking if any of the args possible placements contain Shard(). If they don't, for every argument, we calculate the index of the last Partial placement that doesn't match the preserved reduce_op (e.g., "max" for torch.maximum). This is because for this placement, we will have to redistribute to Replicate, and in a correct implementation, we would have to Replicate all placements that come before it (including Partials that do match the reduce_op). In summary choosing the argument where the last non-matching Partials appear earliest in the placement list minimizes the number of mesh dimensions requiring redistribution. 

**Design Choice/Question**
1. Converted partial_preserving_ops to map to list instead of strings as we can convert linear_ops to follow this format, and some ops can propagate multiple types of partials.

2. Currently I check every arg before hand if it contains a spec with Shard. The reason I can't just use max_shard to tell me this is because the max_num_shards() returns atleast 1 no matter what (even with no sharding) However, there is, what i think is an extreme edge case where if the mesh_size is 1 and there's one shard placement, we return 1. If we accept that in this instance, we're ok with converting to Partial sometimes, i can remove logic checking the arg for Shard and just do if preserve_partial is not None and max_shards == 1:

**Test Cases**
1. pytest test/distributed/tensor/test_pointwise_ops.py -k test_maximum_mixed_partials_redistribution
2. pytest test/distributed/tensor/test_pointwise_ops.py -k test_maximum_selects_optimal_strategy




cc wanchaol tianyu-l wz337 XilunWu d4l3k pragupta SherlockNoMad ppwwyyxx

[ghstack-poisoned]

anshul-si committed:
Update on "[dtensor][partial] improving followed_strategy selection when no sharding placement for min/max"

**Summary:** The current problem is when we have no sharding placements in any of the input_args, we choose first arg's strategy as followed_strategy in pointwise_ops, which can lead to unnecessary redistributes. One example of this is torch.maximum(p(max), p(sum)) vs torch.maximum(p(sum), p(max)). In the first instance, we choose p(max) and propagate it. p(sum) is then converted to replicate. Overall we do one redistribute. However, for the second case, we choose p(sum) to propagate. We choose replicate as linearity is -1, and we check if sum == max. This results in both args being converted to Replicate. 

I solve this issue by first checking if any of the args possible placements contain Shard(). If they don't, for every argument, we calculate the index of the last Partial placement that doesn't match the preserved reduce_op (e.g., "max" for torch.maximum). This is because for this placement, we will have to redistribute to Replicate, and in a correct implementation, we would have to Replicate all placements that come before it (including Partials that do match the reduce_op). In summary choosing the argument where the last non-matching Partials appear earliest in the placement list minimizes the number of mesh dimensions requiring redistribution. 

**Design Choice/Question**
1. Converted partial_preserving_ops to map to list instead of strings as we can convert linear_ops to follow this format, and some ops can propagate multiple types of partials.

2. Currently I check every arg before hand if it contains a spec with Shard. The reason I can't just use max_shard to tell me this is because the max_num_shards() returns atleast 1 no matter what (even with no sharding) However, there is, what i think is an extreme edge case where if the mesh_size is 1 and there's one shard placement, we return 1. If we accept that in this instance, we're ok with converting to Partial sometimes, i can remove logic checking the arg for Shard and just do if preserve_partial is not None and max_shards == 1:

**Test Cases**
1. pytest test/distributed/tensor/test_pointwise_ops.py -k test_maximum_mixed_partials_redistribution
2. pytest test/distributed/tensor/test_pointwise_ops.py -k test_maximum_selects_optimal_strategy




cc wanchaol tianyu-l wz337 XilunWu d4l3k pragupta SherlockNoMad ppwwyyxx

[ghstack-poisoned]

anshul-si committed:
Update on "[dtensor][partial] improving followed_strategy selection when no sharding placement for min/max"

**Summary:** The current problem is when we have no sharding placements in any of the input_args, we choose first arg's strategy as followed_strategy in pointwise_ops, which can lead to unnecessary redistributes. One example of this is torch.maximum(p(max), p(sum)) vs torch.maximum(p(sum), p(max)). In the first instance, we choose p(max) and propagate it. p(sum) is then converted to replicate. Overall we do one redistribute. However, for the second case, we choose p(sum) to propagate. We choose replicate as linearity is -1, and we check if sum == max. This results in both args being converted to Replicate. 

I solve this issue by first checking if any of the args possible placements contain Shard(). If they don't, for every argument, we calculate the index of the last Partial placement that doesn't match the preserved reduce_op (e.g., "max" for torch.maximum). This is because for this placement, we will have to redistribute to Replicate, and in a correct implementation, we would have to Replicate all placements that come before it (including Partials that do match the reduce_op). In summary choosing the argument where the last non-matching Partials appear earliest in the placement list minimizes the number of mesh dimensions requiring redistribution. 

**Design Choice/Question**
1. Converted partial_preserving_ops to map to list instead of strings as we can convert linear_ops to follow this format, and some ops can propagate multiple types of partials.

2. Currently I check every arg before hand if it contains a spec with Shard. The reason I can't just use max_shard to tell me this is because the max_num_shards() returns atleast 1 no matter what (even with no sharding) However, there is, what i think is an extreme edge case where if the mesh_size is 1 and there's one shard placement, we return 1. If we accept that in this instance, we're ok with converting to Partial sometimes, i can remove logic checking the arg for Shard and just do if preserve_partial is not None and max_shards == 1:

**Test Cases**
1. pytest test/distributed/tensor/test_pointwise_ops.py -k test_maximum_mixed_partials_redistribution
2. pytest test/distributed/tensor/test_pointwise_ops.py -k test_maximum_selects_optimal_strategy




cc wanchaol tianyu-l wz337 XilunWu d4l3k pragupta SherlockNoMad ppwwyyxx

[ghstack-poisoned]

anshul-si committed:
Update on "[dtensor][partial] improving followed_strategy selection when no sharding placement for min/max"

**Summary:** The current problem is when we have no sharding placements in any of the input_args, we choose first arg's strategy as followed_strategy in pointwise_ops, which can lead to unnecessary redistributes. One example of this is torch.maximum(p(max), p(sum)) vs torch.maximum(p(sum), p(max)). In the first instance, we choose p(max) and propagate it. p(sum) is then converted to replicate. Overall we do one redistribute. However, for the second case, we choose p(sum) to propagate. We choose replicate as linearity is -1, and we check if sum == max. This results in both args being converted to Replicate. 

I solve this issue by first checking if any of the args possible placements contain Shard(). If they don't, for every argument, we calculate the index of the last Partial placement that doesn't match the preserved reduce_op (e.g., "max" for torch.maximum). This is because for this placement, we will have to redistribute to Replicate, and in a correct implementation, we would have to Replicate all placements that come before it (including Partials that do match the reduce_op). In summary choosing the argument where the last non-matching Partials appear earliest in the placement list minimizes the number of mesh dimensions requiring redistribution. 

**Design Choice/Question**
1. Converted partial_preserving_ops to map to list instead of strings as we can convert linear_ops to follow this format, and some ops can propagate multiple types of partials.

2. Currently I check every arg before hand if it contains a spec with Shard. The reason I can't just use max_shard to tell me this is because the max_num_shards() returns atleast 1 no matter what (even with no sharding) However, there is, what i think is an extreme edge case where if the mesh_size is 1 and there's one shard placement, we return 1. If we accept that in this instance, we're ok with converting to Partial sometimes, i can remove logic checking the arg for Shard and just do if preserve_partial is not None and max_shards == 1:

**Test Cases**
1. pytest test/distributed/tensor/test_pointwise_ops.py -k test_maximum_mixed_partials_redistribution
2. pytest test/distributed/tensor/test_pointwise_ops.py -k test_maximum_selects_optimal_strategy




cc wanchaol tianyu-l wz337 XilunWu d4l3k pragupta SherlockNoMad ppwwyyxx

[ghstack-poisoned]
